### ISSUE TITLE
Why does data loading of the 1st iter cost more time than other iters.

### ISSUE THREAD

**Issue by RaymondByc** [open-mmlab/mmdetection#6020]
Why does data loading of the 1st iter cost more time in each epoch than other iters?  
What happened at this time?

**Comment by jshilong** [comment]
If you use multiple workers, the dataloader will lunch `num_works` subprocesses, which is a time-consuming operation.

**Comment by RaymondByc** [comment]
Why lunch subprocesses in every epoch? Is that necessary? 

We concern about this issue because the official code of YOLOX does not this time cost. 

Here is the training log:

-------------------------------------------------------------------------------------------------------------------------------------------
Log of the official code: 


2021-09-02 01:36:17.800 | INFO     | yolox.core.trainer:after_iter:246 - epoch: 1/300, iter: 1840/1849, mem: 7420Mb, iter_time: 0.404s, data_time: 0.025s, total_loss: 11.3, iou_loss: 3.5, l1_loss: 0.0, conf_loss: 5.0, cls_loss: 2.9, lr: 3.961e-04, size: 640, ETA: 2 days, 8:39:37
2021-09-02 01:36:20.945 | INFO     | yolox.core.trainer:save_ckpt:318 - Save weights to ./YOLOX_outputs/yolox_s
2021-09-02 01:36:21.124 | INFO     | yolox.core.trainer:before_epoch:188 - ---> start train epoch2
2021-09-02 01:36:24.208 | INFO     | yolox.core.trainer:after_iter:246 - epoch: 2/300, iter: 10/1849, mem: 7420Mb, iter_time: 0.307s, data_time: 0.009s, total_loss: 11.0, iou_loss: 3.5, l1_loss: 0.0, conf_loss: 4.8, cls_loss: 2.7, lr: 4.043e-04, size: 544, ETA: 2 days, 8:35:31

----------------------------------------------------------------------------------------------------------------------------------------------
Log of the mmdet code: 

2021-09-01 02:51:31,651 - mmdet - INFO - Epoch [1][1800/1849]	lr: 3.791e-04, eta: 3 days, 20:39:50, time: 0.488, data_time: 0.025, memory: 4798, loss_cls: 2.6535, loss_bbox: 3.5730, loss_obj: 5.2453, loss: 11.4718
2021-09-01 02:56:24,645 - mmdet - INFO - Epoch [2][50/1849]	lr: 4.219e-04, eta: 4 days, 12:01:20, time: 4.997, data_time: 4.050, memory: 7597, loss_cls: 2.6952, loss_bbox: 3.5113, loss_obj: 5.7309, loss: 11.9374

**Comment by qq1628qq** [comment]
why my time>4, data_time: >3?and train time is so long .

**Comment by RangiLyu** [comment]
> Why lunch subprocesses in every epoch? Is that necessary?
> 


This is the original design of the Dataloader in pytorch<1.9. Official YOLOX does not need to relaunch subprocesses because they implemented `InfiniteSampler` to avoid this problem. We also implemented an `InfiniteSampler` in this PR #5996 but not merge yet.

**Comment by RaymondByc** [comment]
I test PR #5996 in my machine, it cannot achieve the training speed of the official code, even though it is faster than epoch_based_runner versions. 

Even though it reduces the time of switching epochs, the time on switching iterations is added.

Any solutions to resolve this problem?

**Comment by jshilong** [comment]
> I test PR #5996 in my machine, it cannot achieve the training speed of the official code, even though it is faster than epoch_based_runner versions.
> 
> Even though it reduces the time of switching epochs, the time on switching iterations is added.
> 
> Any solutions to resolve this problem?

@hhaAndroid Please take a look at this issue

**Comment by RaymondByc** [comment]
Any idea?

**Comment by jshilong** [comment]
Would you mind explaining the difference with the official repo？ @hhaAndroid

**Comment by ZwwWayne** [comment]
> > Why lunch subprocesses in every epoch? Is that necessary?
> 
> This is the original design of the Dataloader in pytorch<1.9. Official YOLOX does not need to relaunch subprocesses because they implemented `InfiniteSampler` to avoid this problem. We also implemented an `InfiniteSampler` in this PR #5996 but not merge yet.

This reply already answers the question so this issue will be closed. For iteration-based training, we do not have time to try it yet but we will do it in the following months. Please @RaymondByc create a new issue to track your question. We may fix some potential bugs/differences if we reached a conclusion.
